src/solver/simulation/include/antares/solver/simulation/solver.h

```diff
@@ -34,6 +34,40 @@
 #include "antares/solver/simulation/solver_utils.h"
 #include "antares/solver/variable/state.h"
 
+class NumSpaceManager
+{
+public:
+    NumSpaceManager(int N):
+        available(N, true)
+    {
+    }
+
+    inline int getAvailableNumSpace()
+    {
+        // std::find not available for std::vector<bool>
+        std::unique_lock lk(mut);
+        int idx = 0;
+        for (int idx = 0; idx < available.size(); idx++)
+        {
+            if (available[idx])
+            {
+                available[idx] = false;
+                return idx;
+            }
+        }
+        return -1;
+    }
+
+    inline void freeNumSpace(int numSpace)
+    {
+        std::unique_lock lk(mut);
+        available[numSpace] = true;
+    }
+
+    std::mutex mut;
+    std::vector<bool> available;
+};
+
```

Comment:
```suggestion
#include <boost/lockfree/queue.hpp>

class NumSpaceManager
{
public:
    NumSpaceManager(int N) : availableSpaces(N), maxSize(N)
    {
        for (int i = 0; i < N; ++i)
        {
            availableSpaces.push(i);
        }
    }

    int getAvailableNumSpace()
    {
        int idx;
        return availableSpaces.pop(idx) ? idx : -1;
    }

    void freeNumSpace(int numSpace)
    {
    // check against maxSize
        while (!availableSpaces.push(numSpace)); // Ensure push succeeds
    }

private:
    boost::lockfree::queue<int> availableSpaces;
    const int maxSize;
};
```

Reply:
By design the mutex doesn't lock, and I find the current implementation more clear, except it should throw, not return -1.